core/src/worker/activities.rs

```diff
@@ -509,6 +539,72 @@ where
+                                        }
+                                    })
+                                    .zip(task.resp.scheduled_time.clone())
+                                    .and_then(|(to, st)| {
+                                        let now = SystemTime::now();
+                                        let sched_time = SystemTime::try_from(st).unwrap_or(now);
+                                        let already_elapsed =
+                                            now.duration_since(sched_time).unwrap_or_default();
+                                        Duration::try_from(to).map(|to| to - already_elapsed).ok()
+                                    });
+                                let timeout_at = [
+                                    task.resp.heartbeat_timeout.clone(),
+                                    task.resp.start_to_close_timeout.clone(),
+                                ]
+                                .into_iter()
+                                .enumerate()
+                                .filter_map(|(i, d)| {
+                                    d.and_then(|d| Duration::try_from(d).ok().map(|d| (i, d)))
+                                })
+                                .chain(sched.map(|s| (2, s)))
+                                .filter(|(_, d)| !d.is_zero())
+                                .min_by(|(_, d1), (_, d2)| d1.cmp(d2));
+                                if let Some((timeout_type, timeout_at)) = timeout_at {
+                                    let sleep_time = timeout_at + local_timeout_buffer;
+                                    let cancel_tx = cancels_tx.clone();
+                                    let resetter = if timeout_type == 0 {
+                                        Some(Arc::new(Notify::new()))
+                                    } else {
+                                        None
+                                    };
+                                    let resetter_clone = resetter.clone();
+                                    outstanding_info.local_timeouts_task =
+                                        Some(tokio::task::spawn(async move {
+                                            if let Some(rs) = resetter_clone {
+                                                loop {
+                                                    tokio::select! {
+                                                        _ = rs.notified() => continue,
+                                                        _ = tokio::time::sleep(sleep_time) => break,
+                                                    }
+                                                }
+                                            } else {
+                                                tokio::time::sleep(sleep_time).await;
+                                            }
+                                            let _ = cancel_tx.send(PendingActivityCancel {
```

Comment:
Do we want to increment a metric here or log or somehow let lang know via the proto that this was due to local timeout? I'm just starting to get scared that a user's clock is gonna skew like 10s off of server's and things are gonna just start cancelling everywhere.

Or I guess we can look at all the activity failures and if they are failures for untriggered cancel we can assume it was either worker shutdown or this timeout? That's probably good enough.

Reply:
I suppose skew is only an issue with the scheduled timeout, since that's the only one that uses an absolute time. We could remove it for just that reason, since you're more likely to fail an attempt timeout or heartbeat anyway.

I don't want to add another metric... we've already got so many. I'll add a debug log line though.

Reply:
:+1: Very happy to see schedule timeout removed. I think it's much safer to do single-attempt duration than clock-based. May be worth checking whether server _always_ sets start-to-close, I forget, but I do feel safer now w/ what is here.